### PR TITLE
WT-7830 Migrate the python setup scripts to use cmake

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by quchenhao** [comment]
@ddanderson @alisonfel I tested this on ubuntu, redhat, and macos. It all works fine. Looks like originally we don't support windows so I don't bother to test there.

File: lang/python/setup_pip.py
Comment (ddanderson) [inline lang/python/setup_pip.py:304]: Would it be clearer to `shutil.copytree` the whole tree, and then selectively rename the `init.py` file?

File: lang/python/setup_pip.py
Comment (ddanderson) [inline lang/python/setup_pip.py:250]: I think we need to put a dependency on cmake and ninja to make sure they are installed.  Some info how to do that here: https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html .  Looks like setting `install_requires` in the `setup()` function call will do the trick.  Would be good to test this on a system that didn't have ninja and/or cmake installed.

**Review by ddanderson — COMMENTED** [review]
@quchenhao Thanks, this is a nice step forward.

I tagged a requirements issue for cmake/ninja.  The other comment is only a suggestion.
Everything else LGTM.

**Review by alisonfel — COMMENTED** [review]
I'm a bit uncertain the purpose of `setup.py` and when it should be run. Though I think we also need to update the path `wiredtiger_wrap.c` to the CMake generated wrap file: `CMakeFiles/__wiredtiger.dir/wiredtigerPYTHON_wrap.c`

https://github.com/wiredtiger/wiredtiger/blob/c6aed282c93fb8f01634118e4e53db953489e666/lang/python/setup.py#L51-L59

I'm assuming `setup.py` usually run after a build step? Since that's when the `wiredtigerPYTHON_wrap.c` gets generated.

**Comment by quchenhao** [comment]
> I'm a bit uncertain the purpose of `setup.py` and when it should be run. Though I think we also need to update the path `wiredtiger_wrap.c` to the CMake generated wrap file: `CMakeFiles/__wiredtiger.dir/wiredtigerPYTHON_wrap.c`
> 
> https://github.com/wiredtiger/wiredtiger/blob/c6aed282c93fb8f01634118e4e53db953489e666/lang/python/setup.py#L51-L59
> 
> I'm assuming `setup.py` usually run after a build step? Since that's when the `wiredtigerPYTHON_wrap.c` gets generated.

I think it is run at the install time. I'll update the path.

File: lang/python/setup.py
Comment (quchenhao) [inline lang/python/setup.py:53]: @ddanderson I don't know how to fix this. I just realized that auto tools actually depends on this:

```
_wiredtiger.so: $(top_builddir)/libwiredtiger.la $(PYSRC)/wiredtiger_wrap.c
	(cd $(PYSRC) && \
	    $(PYTHON) setup.py build_ext -f -b $(abs_builddir) $(PYDIRS))
```

Comment (quchenhao) [reply lang/python/setup.py:53]: @alisonfel I cannot update the path in setup.py. Not sure what I can do here.

File: lang/python/setup_pip.py
Comment (quchenhao) [reply lang/python/setup_pip.py:248]: I tried to add cmake ninja to the python dependency but that is not working. I think the problem is that we install cmake and ninja in python. But when we build we run a shell command in python that doesn't look into the python package so cmake is still not found.

Comment (quchenhao) [reply lang/python/setup_pip.py:221]: I think this is not working because cmake is using a different way to locate external libraries. If I use cmake installed by python, it seems not finding other dependencies even I set LD_LIBRARY_PATH. @alisonfel

**Comment by quchenhao** [comment]
@alisonfel I'll merge this after you merge your fix for cmake.

File: lang/python/setup_pip.py
Comment (alisonfel) [inline lang/python/setup_pip.py:287]: If `os.getenv`can't find the env variable it will default return `None`, which I think will fail when passing it into `call()` . You should be able override the default value though with an empty string e.g.: `os.getenv('CMAKE_LIBRARY_PATH', default='')` 
https://docs.python.org/3/library/os.html#os.getenv

File: lang/python/setup.py
Comment (alisonfel) [reply lang/python/setup.py]: @quchenhao  What type of error occurs when you try to include the path in `setup.py`?

**Comment by alisonfel** [comment]
> @alisonfel I'll merge this after you merge your fix for cmake.

Yep! The PR is up #6839. Would you be able to approve?

File: lang/python/setup_pip.py
Comment (alisonfel) [reply lang/python/setup_pip.py:287]: For context, run I run the script on my mac, it crashes due to a `None` value in the build environment.

**Comment by ddanderson** [comment]
@quchenhao, I wanted to try out this branch locally, but I get this:
```
$ python setup_pip.py sdist
Traceback (most recent call last):
  File "setup_pip.py", line 268, in <module>
    shutil.copytree(pywt_path, stage_dir, dirs_exist_ok=True)
TypeError: copytree() got an unexpected keyword argument 'dirs_exist_ok'
$ python --version
Python 3.6.9
```
My main concern, and why I wanted to test this, is that if `cmake` or `ninja` is not installed, that we do something reasonable to steer the user to install it themselves.  However, that may not be easy for a user that only knows about Python and may not have privileges on the machine they are installing in.  The best case would be to install cmake/ninja via a python dependency.  I know you looked at that some, is there any way to resolve that?

**Comment by quchenhao** [comment]
> @quchenhao, I wanted to try out this branch locally, but I get this:
> 
> ```
> $ python setup_pip.py sdist
> Traceback (most recent call last):
>   File "setup_pip.py", line 268, in <module>
>     shutil.copytree(pywt_path, stage_dir, dirs_exist_ok=True)
> TypeError: copytree() got an unexpected keyword argument 'dirs_exist_ok'
> $ python --version
> Python 3.6.9
> ```
> 
> My main concern, and why I wanted to test this, is that if `cmake` or `ninja` is not installed, that we do something reasonable to steer the user to install it themselves. However, that may not be easy for a user that only knows about Python and may not have privileges on the machine they are installing in. The best case would be to install cmake/ninja via a python dependency. I know you looked at that some, is there any way to resolve that?

The cmake and ninja dependency is resolved and they will be automatically installed by the script. It looks like the flag is python 3.9 specific. I'll check if there is any other choice.

File: lang/python/setup_pip.py
Comment (quchenhao) [reply lang/python/setup_pip.py:287]: Thanks for the tip.

File: lang/python/setup_pip.py
Comment (quchenhao) [inline lang/python/setup_pip.py:352]: This installs cmake and ninja automatically.

File: lang/python/setup_pip.py
Comment (quchenhao) [inline lang/python/setup_pip.py:267]: copytree can only work for us from python 3.8+. Revert back to the old way of copying directory.

File: lang/python/setup_pip.py
Comment (quchenhao) [reply lang/python/setup_pip.py]: Fixed and tested.

**Review by alisonfel — APPROVED** [review]
LGTM

**Comment by ddanderson** [comment]
@quchenhao and @alisonfel, I'm still having some trouble with this when I try to actually use it.  On the develop branch I can do this:
```
(py3) $ python -mvenv xxx
(py3) $ . ./xxx/bin/activate
(xxx) $ python -mpip install ./dist/wiredtiger-10.0.1.tar.gz 
Looking in indexes: https://artifactory.corp.mongodb.com/artifactory/api/pypi/pypi/simple, https://pypi.org/simple
Processing ./dist/wiredtiger-10.0.1.tar.gz
Building wheels for collected packages: wiredtiger
  Building wheel for wiredtiger (setup.py) ... done
  Created wheel for wiredtiger: filename=wiredtiger-10.0.1-cp36-cp36m-linux_x86_64.whl size=1095860 sha256=f581fb7b675fe62c35d6a53b04a1b4f1901994b8384cd1555ab758d5fa3d057c
  Stored in directory: /home/ubuntu/.cache/pip/wheels/23/41/9f/7ea6f6e2c2f02bf3b48d6c57149db7f6b966caa6cca99f25d6
Successfully built wiredtiger
Installing collected packages: wiredtiger
  Attempting uninstall: wiredtiger
    Found existing installation: wiredtiger 10.0.1
    Uninstalling wiredtiger-10.0.1:
      Successfully uninstalled wiredtiger-10.0.1
Successfully installed wiredtiger-10.0.1
WARNING: You are using pip version 20.2.2; however, version 21.2.1 is available.
You should consider upgrading via the '/home/ubuntu/wt/git/develop/lang/python/xxx/bin/python -m pip install --upgrade pip' command.
(xxx) $ cd ../../
(xxx) $ python test/py_install/testbase.py 
Expect 200 = 200
['WiredTiger 10.0.1: (April 12, 2021)', 10, 0, 1]
testbase success.
(xxx) $ python test/py_install/test
testbase.py  testpack.py  
(xxx) $ python test/py_install/testpack.py 
testpack success.
(xxx) $
```
But with this branch, 
```
(xxx) $ python test/py_install/testbase.py 
Traceback (most recent call last):
  File "test/py_install/testbase.py", line 31, in <module>
    import wiredtiger, shutil, os
  File "/home/ubuntu/mongo/py3/lib/python3.6/site-packages/wiredtiger/__init__.py", line 52, in <module>
    import _wiredtiger
ImportError: /home/ubuntu/mongo/py3/lib/python3.6/site-packages/wiredtiger/_wiredtiger.cpython-36m-x86_64-linux-gnu.so: undefined symbol: __wt_calloc
(xxx) $
```
Are you seeing the same thing?  Is there an environment variable that needs to be set?  It would be nice for it to work out the box.  Let me know if you want me to take a closer look at it.

**Review by ddanderson — COMMENTED** [review]
I put my review as a regular comment.  Hoping to see that this change can be installed and run simple programs.

File: lang/python/setup_pip.py
Comment (quchenhao) [inline lang/python/setup_pip.py:229]: @alisonfel Do we still need this line?

File: lang/python/setup_pip.py
Comment (quchenhao) [inline lang/python/setup_pip.py]: @alisonfel Why we remove this line? We don't build this anymore?

File: lang/python/setup_pip.py
Comment (alisonfel) [reply lang/python/setup_pip.py:229]: Yep, we still need this line. We need to invoke the python build to generate the SWIG files, which in turn are needed to compile the python library.

File: lang/python/setup_pip.py
Comment (alisonfel) [reply lang/python/setup_pip.py]: I don't believe we need this line. That change reverts back to relying on the setup process building the shared python API library (as it was doing prior to our changes). Trying to bypass the `build_ext` process by dropping in our own binary seemed to cause more issues and also doesn't stop it from compiling its own python `_wiredtiger` library.

I think we should only need the CMake build to generate the libwiredtiger archive and the SWIG files (which was how it was done in the autoconf build)

**Comment by alisonfel** [comment]
I've also scheduled the Evergreen CMake tests since this PR impacts the build system.

File: CMakeLists.txt
Comment (ddanderson) [inline CMakeLists.txt:103]: depenends => depends.

File: lang/python/setup_pip.py
Comment (ddanderson) [inline lang/python/setup_pip.py:283]: "As a temporary solution" -- I'd remove this phrase.  Since the fix for this is in CMake, we don't know if it's temporary or if it will ever be done.  You could say "As a workaround", but then I would put the second sentence as the first sentence.

**Review by ddanderson — APPROVED** [review]
LGTM -- I suggested some small comment improvements.
Nicely done.

File: CMakeLists.txt
Comment (alisonfel) [reply CMakeLists.txt]: Good pickup. Fixed!

File: lang/python/setup_pip.py
Comment (alisonfel) [reply lang/python/setup_pip.py]: Good point. I reworded the comment per your suggestion. I also added the ticket that will fix the CMake builtin support (tracked as a FIXME)

**Review by lukech — APPROVED** [review]
Assuming we need to build using cmake before calling `setup_pip.py`, can you advise the needed update to the [Python installation wiki page](https://github.com/wiredtiger/wiredtiger/wiki/HOWTO:UploadAReleaseForPythonInstalls), specifically the "Preparing a release for upload" section?

**Comment by quchenhao** [comment]
> Assuming we need to build using cmake before calling `setup_pip.py`, can you advise the needed update to the [Python installation wiki page](https://github.com/wiredtiger/wiredtiger/wiki/HOWTO:UploadAReleaseForPythonInstalls), specifically the "Preparing a release for upload" section?

No, we don't need to do that.

**Pull request merged**
